Lib/socket.py

```diff
@@ -728,6 +728,140 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+                  reuse_port=False, flags=None, hybrid_ipv46=False):
+    """Convenience function which creates a socket bound to *address*
+    (a 2-tuple (host, port)) and return the socket object.
+
+    If *host* is an empty string or None all network interfaces are assumed.
+
+    If *family* is AF_UNSPEC the address family will be determined from the
+    *host* specified in *address* and AF_INET will take precedence over
+    AF_INET6 in case family can't be determined from *host*.
+
+    *type* should be either SOCK_STREAM or SOCK_DGRAM.
+
+    *backlog* is the queue size passed to socket.listen() and is ignored
+    for SOCK_DGRAM socket types.
+
+    *reuse_port* dictate whether to use the SO_REUSEPORT socket option.
+
+    *flags* is a bitmask for getaddrinfo().
+
+    *hybrid_ipv46*: if True and the platform supports it it will create
+    a socket able to accept both IPv4 and IPv6 connections.
+    When False it will explicitly disable this option on platforms that
+    enable it by default (e.g. Linux).
+
+    >>> with create_server((None, 8000)) as server:
+    ...     while True:
+    ...         conn, addr = server.accept()
+    ...         # handle new connection
+    """
+    host, port = address
+    if host == "":
+        # When using getaddrinfo(), None is an alias for "all interfaces":
+        # https://mail.python.org/pipermail/python-ideas/2013-March/019937.html
+        host = None  # all interfaces
+    # Note about Windows: we don't set SO_REUSEADDR because:
+    # 1) It's unnecessary: bind() will succeed even in case of a
+    # previous closed socket on the same address and still in TIME_WAIT
+    # state.
+    # 2) If set, another socket will be free to bind() on the same
+    # address, effectively preventing this one from accepting connections.
+    # Also, it sets the process in a state where it'll no longer respond
+    # to any signals or graceful kills. More at:
+    # See: msdn2.microsoft.com/en-us/library/ms740621(VS.85).aspx
+    reuse_addr = os.name not in ('nt', 'cygwin') and \
+        hasattr(_socket, 'SO_REUSEADDR')
+    if reuse_port and not hasattr(_socket, "SO_REUSEPORT"):
+        raise ValueError("SO_REUSEPORT not supported on this platform")
+    if type not in (SOCK_STREAM, SOCK_DGRAM):
+        raise ValueError("only SOCK_STREAM and SOCK_DGRAM types are supported")
+    if flags is None:
+        flags = AI_PASSIVE
+    if hybrid_ipv46:
+        if type != SOCK_STREAM:
+            raise ValueError("hybrid_ipv46 requires SOCK_STREAM type")
+        if family not in {AF_INET6, AF_UNSPEC}:
+            raise ValueError("hybrid_ipv46 requires AF_INET6 family")
+        if not supports_hybrid_ipv46():
+            raise ValueError("hybrid_ipv46 not supported on this platform")
+        family = AF_INET6
+    info = getaddrinfo(host, port, family, type, 0, flags)
+    if family == AF_UNSPEC:
+        # prefer AF_INET over AF_INET6
```

Comment:
I'm strongly opposed to this choice. If the IP stack is giving preference to IPv6 by presenting it first, Python should not be overriding it to prefer the less capable legacy approach. I see this approach as being comparable to preferring Python 2 syntax in a compatibility library. The systems should be designed to prefer the newer, more advanced behavior over the older one, even if the older is dominant. Giving preference to IPv4 leaves technical debt that eventually will need to be rectified (with all of the transitional problems that brings).

In other words, I believe this choice precludes the ability to solve my primary use-case for this function, which is:

"Create a socket listening on the best available protocol" where best available is IPv6+IPv4 or IPv6 or IPv4 in descending order.

Reply:
The use case you're describing is already covered by:
```python
create_server(addr, hybrid_ipv46=support_hybrid_ipv46())
```

Reply:
Okay. I'd like to run some tests.

Reply:
Okay. I confirm this usage behaves the way I would expect. I suppose I could accept this approach, if necessary.

I'd have preferred this behavior would be the default, as that's the default server behavior I've used on other IP servers I've written. I find it awkward that the default is "prefer the legacy behavior", especially for a new function for which there's no prior expectation. At what point do you think Python would want to change this default behavior to prefer IPv6?